### PR TITLE
Rename `Magic*` to `IpyEscape*`

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR renames the `MagicCommand` token to `IpyEscapeCommand` token and `MagicKind` to `IpyEscapeKind` type to better reflect the purpose of the token and type. Similarly, it renames the AST nodes from `LineMagic` to `IpyEscapeCommand` prefixed with `Stmt`/`Expr` wherever necessary.

It also makes renames from using `jupyter_magic` to `ipython_escape_commands` in various function names.

The mode value is still `Mode::Jupyter` because the escape commands are part of the IPython syntax but the lexing/parsing is done for a Jupyter notebook.

### Motivation behind the rename:
* IPython codebase defines it as "EscapeCommand" / "Escape Sequences":
    * Escape Sequences: https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L329-L333
    * Escape command: https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L410-L411
* The word "magic" is used mainly for the actual magic commands i.e., the ones starting with `%`/`%%` (https://ipython.readthedocs.io/en/stable/interactive/reference.html#magic-command-system). So, this avoids any confusion between the Magic token (`%`, `%%`) and the escape command itself.
## Test Plan

* `cargo test` to make sure all renames are done correctly.
* `grep` for `jupyter_escape`/`magic` to make sure all renames are done correctly.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6395** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6395" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/6395?utm_source=stack-comment).

**Comment by MichaReiser** [comment]
You mention that the rename is to better reflect the purpose and type. Can you expand on what makes the new name the better choice?

**Comment by dhruvmanila** [comment]
## IPython Escape Commands Terminology

Before we get into the definitions, I'm omitting all of the rules of lexing/parsing because we're only interested in the definition.

* "Magic commands" are special IPython syntax which starts with a token to identify the magic kind followed by the command value itself.
* "Magic kind" are the kind of magic commands which are recognized by the token itself: `%`, `%%`, `!`, `!!`, `?`, `??`, `/`, `;`, and `,`.
* "Help command" (or "Dynamic Object Introspection" as it's called) are the escape commands of the kind `?` and `??`. (`?str.replace`)
* "Help end command" are a subset of "Help command" where the token can be at the end of the line i.e., after the value itself (`str.replace?`). Here's where things get tricky. I'll divide the help end command into two types for better understanding:
    * Strict version: The token is _only_ at the end of the line (`str.replace?`, `str.replace??`)
    * Combined version: Not only either `?` or `??` is at the end of the line but other magic kind tokens are present at the start as well (`%matplotlib?`, `%%timeit?`)
* "Priority" comes into picture for the "Combined version" mentioned above. When there are magic kind tokens on both side of the value, how do we determine the magic kind i.e., which token to choose? The "Help end command" always takes priority over any other token i.e., if there is `?`/`??` at the end then that is used to determine the kind.

### What the syntax would look like?
* `<MagicKind><Command value>` (e.g., `%matplotlib`, `/foo`, `!pwd`, etc.)
* `<Command value><MagicKind ("?" or "??")>` (e.g., `str.replace?`, `math.pi??`, etc.)
* `<MagicKind><Command value><MagicKind ("?" or "??")>` (e.g., `%matplotlib?`, `%%timeit??`, etc.)

### Motivation behind the rename:
* IPython codebase defines it as "EscapeCommand" / "Escape Sequences":
    * Escape Sequences: https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L329-L333
    * Escape command: https://github.com/ipython/ipython/blob/292e3a23459ca965b8c1bfe2c3707044c510209a/IPython/core/inputtransformer2.py#L410-L411
* The word "magic" is used mainly for the actual magic commands i.e., the ones starting with `%`/`%%` (https://ipython.readthedocs.io/en/stable/interactive/reference.html#magic-command-system). So, this avoids any confusion between the Magic token (`%`, `%%`) and the escape command itself.

**Comment by MichaReiser** [comment]
Thanks for the explanation. This is very helpful context that I didn't have before. We should preserve this in writing somewhere (maybe as documentation on `EscapeCommand`). 

A few thoughts:

> The word "magic" is used mainly for the actual magic commands i.e., the ones starting with %/%% 

That makes sense now. So only the EscapeKind `%` and `%%` are "magic"

What I liked about the old "Magic" prefix is that it clearly marked them as not being part of the Python spec. Would it make sense to keep a similar prefix. Maybe IpyEscapeCommand and IpyEscapeSequence. Should we use the same prefix for the tokens and ast nodes? 

We did something similar at rome where we prefixed the JS Spec nodes with `Js`, e.g. `JsBinaryExpression`, the type script syntax nodes with `Ts` (`TsTypeAliasDeclaration`), and the jsx nodes with `Jsx` (`JsxTagExpression`)

**Comment by MichaReiser** [comment]
Sounds good to me. 

> StmtLineMagic/ExprLineMagic -> IpyEscapeCommandStatement/IpyEscapeCommandExpression (Is this too verbose?) or StmtIpyEscapeCommand/ExprIpyEscapeCommand (for consistency with the current naming convention).

I prefer the `IpyEscapeCommandStatement` but we may need to go with `StmtIpyEscapeCommand` for consistency (until we rename all nodes)

File: crates/ruff_python_ast/src/node.rs [hunk 1]
```diff
@@ -48,7 +48,7 @@ pub enum AnyNode {
     StmtPass(ast::StmtPass),
     StmtBreak(ast::StmtBreak),
     StmtContinue(ast::StmtContinue),
-    StmtLineMagic(ast::StmtLineMagic),
+    StmtLineMagic(ast::StmtIpyEscapeCommand),
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/node.rs:51]: Rename the node type

File: crates/ruff_python_ast/src/node.rs [hunk 2]
```diff
@@ -76,7 +76,7 @@ pub enum AnyNode {
     ExprList(ast::ExprList),
     ExprTuple(ast::ExprTuple),
     ExprSlice(ast::ExprSlice),
-    ExprLineMagic(ast::ExprLineMagic),
+    ExprLineMagic(ast::ExprIpyEscapeCommand),
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/node.rs:79]: Rename the variant name

File: crates/ruff_python_ast/src/comparable.rs [hunk 3]
```diff
@@ -1166,7 +1166,7 @@ pub struct StmtExpr<'a> {
 
 #[derive(Debug, PartialEq, Eq, Hash)]
 pub struct StmtLineMagic<'a> {
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/comparable.rs:1168]: Rename the struct name

**Review by MichaReiser — APPROVED** [review]
Looks good. You may want to search for `StmtLineMagic` and `ExprLineMagic` (text) and replace all occurrences with the new names

**Comment by dhruvmanila** [comment]
Thanks for noticing the remaining renames! :)

**Pull request merged**